### PR TITLE
Open liberty Server instrumentations

### PR DESCRIPTION
**Pull request by kevinfwl**
Added instrumentation for open liberty servers. Please let me know if I should add more snapshot test cases to make the instrumentation more robust.

### CODE REVIEW

File: dd-java-agent/instrumentation/liberty-20/liberty-20.gradle [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+apply plugin: 'org.unbroken-dome.test-sets'
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  //    ibm liberty libraries that were tried but didn't work
```
Comment (kevinfwl) [inline dd-java-agent/instrumentation/liberty-20/liberty-20.gradle:13]: @DataDog/apm-java Unfortunately, I was not able to find the open liberty jar files as artifacts on maven, and consequently did not include a muzzle for this instrumentation. Is there a more ideal way to handle this scenario?

File: dd-smoke-tests/springboot-openliberty/snapshots/datadog.smoketest.SpringBootOpenLibertySnapshotTest.simple.snap [hunk 2]
```diff
@@ -34,13 +34,13 @@
...
-       "duration" 74330915
+       "start" 1618011719625753238
+       "duration" 76667015
        "meta" {"env" "smoketest"
                "language" "jvm"
                "version" "99"
-               "thread.name" "Default Executor-thread-4"
-               "thread.id" "48"
+               "thread.name" "Default Executor-thread-2"
+               "thread.id" "44"
```
Comment (tylerbenson) [inline dd-smoke-tests/springboot-openliberty/snapshots/datadog.smoketest.SpringBootOpenLibertySnapshotTest.simple.snap:43]: There are a lot of changes here that seem unnecessary as the value will change every time it runs.  How do we make the test harness ignore the resulting value here?

File: dd-java-agent/instrumentation/liberty-20/liberty-20.gradle [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  //    //https://mvnrepository.com/artifact/com.ibm.websphere.appserver.spi/com.ibm.websphere.appserver.spi.containerServices
+  //    compileOnly group: 'com.ibm.websphere.appserver.spi', name: 'com.ibm.websphere.appserver.spi.containerServices', version: '4.0.50'
+  //    // https://mvnrepository.com/artifact/com.ibm.websphere.appserver.spi/com.ibm.websphere.appserver.spi.servlet
+  //    compileOnly group: 'com.ibm.websphere.appserver.spi', name: 'com.ibm.websphere.appserver.spi.servlet', version: '2.7.50'
+  //    // https://mvnrepository.com/artifact/com.ibm.websphere.appserver.spi/com.ibm.websphere.appserver.spi.httptransport
+  //    compileOnly group: 'com.ibm.websphere.appserver.spi', name: 'com.ibm.websphere.appserver.spi.httptransport', version: '4.1.50'
+  //    compileOnly group: 'io.openliberty', name: 'openliberty-runtime', version: '21.0.0.3'
+
+  //TODO: ideally change this to a maven repository instead of using local jars as a compile stub
+  compileOnly fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/liberty-20/liberty-20.gradle:27]: really not loving this.  What do these jars actually contain?  Are they stubs or full downloads from somewhere? There might be some licensing issues adding them here.

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+  public static class HandleRequestAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static AgentScope onEnter(@Advice.Argument(value = 0) ServletRequest req) {
+      if (!(req instanceof SRTServletRequest)) return null;
+      SRTServletRequest request = (SRTServletRequest) req;
+
+      // if we try to get an attribute that doesn't exist open liberty might complain with an
+      // exception
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java:72]: wow... it really throws an exception if the attribute is unset?  That's silly.

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.Enumeration;
+import javax.servlet.http.HttpServletRequest;
+
+public class RequestExtractAdapter implements AgentPropagation.ContextVisitor<HttpServletRequest> {
+
+  public static final RequestExtractAdapter GETTER = new RequestExtractAdapter();
+
+  @Override
+  public void forEachKey(HttpServletRequest carrier, AgentPropagation.KeyClassifier classifier) {
+    Enumeration<String> headerNames = carrier.getHeaderNames();
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java:13]: Does `SRTServletRequest` have any better methods of accessing the headers that might be more efficient?

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/ResponseFinishInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("com.ibm.ws.webcontainer.srt.SRTServletResponse");
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    return singletonMap(
+        named("finish").and(takesNoArguments()),
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/ResponseFinishInstrumentation.java:42]: When is this `finish` called and is it possible for the status code to change afterwards?

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      final AgentSpan.Context.Extracted extractedContext = propagate().extract(request, GETTER);
+      request.setAttribute(DD_EXTRACTED_CONTEXT_ATTRIBUTE, extractedContext);
+      final AgentSpan span =
+          AgentTracer.startSpan(SERVLET_REQUEST, extractedContext).setMeasured(true);
+
+      DECORATE.afterStart(span);
+      DECORATE.onRequest(span, request, request, extractedContext);
+      final AgentScope scope = activateSpan(span);
+      scope.setAsyncPropagation(true);
+      request.setAttribute(DD_SPAN_ATTRIBUTE, span);
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java:90]: I assume that this is preventing the servlet instrumentation from activating... Do your tests verify that?

File: dd-smoke-tests/springboot-openliberty/snapshots/datadog.smoketest.SpringBootOpenLibertySnapshotTest.simple.snap (hunk 2 above)
Comment (kevinfwl) [reply dd-smoke-tests/springboot-openliberty/snapshots/datadog.smoketest.SpringBootOpenLibertySnapshotTest.simple.snap:43]: There is an ignored keys param `snapshot` function in `AbstractTestAgentSmokeTest.groovy` , where an array of keys to be ignored can be passed

File: dd-java-agent/instrumentation/liberty-20/liberty-20.gradle (hunk 3 above)
Comment (kevinfwl) [reply dd-java-agent/instrumentation/liberty-20/liberty-20.gradle:27]: These contain the code to run an embedded open liberty server. I found them by extracting the packaged `springboot-openliberty` jar file (also tried looking through here https://openliberty.io/downloads/). Unfortunately, I could not find them anywhere. Is there a better way to go about this?

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/ResponseFinishInstrumentation.java (hunk 6 above)
Comment (kevinfwl) [reply dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/ResponseFinishInstrumentation.java:42]: From my understanding, this is more or less the final step. At this step, an error report should've been generated by open liberty so I don't believe that the status code would change.

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java (hunk 7 above)
Comment (kevinfwl) [reply dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java:90]: hmm, the tests verified that the span structures are the keep the same and that the component names have changed from `java-web-servlet` to `liberty-server` (see [link](dd-smoke-tests/springboot-openliberty/snapshots/datadog.smoketest.SpringBootOpenLibertySnapshotTest.simple.snap)). Is that proof that the servlet instrumentation is not activating?

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java (hunk 5 above)
Comment (kevinfwl) [reply dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java:13]: Will take a look at this

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java (hunk 7 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/LibertyServerInstrumentation.java:90]: Yeah, that sounds fine to me.

File: dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java (hunk 5 above)
Comment (kevinfwl) [reply dd-java-agent/instrumentation/liberty-20/src/main/java/datadog/trace/instrumentation/liberty20/RequestExtractAdapter.java:13]: After further looking, I think they only have implemented this method

File: dd-java-agent/instrumentation/liberty-20/liberty-20.gradle (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/liberty-20/liberty-20.gradle:27]: Yeah, unfortunately I don't have any better ideas.  Pretty annoying that they're not doing it the right way.

**Review by tylerbenson — APPROVED** [review]
Really unfortunate that IBM's gotta make it so difficult.

**Comment by kevinfwl** [comment]
After further looking and guidance from Brian, it seems that the jars are actually inside the runtime environment. Sorry about the carelessness earlier. I've added some extra configs to Gradle to get the artifacts properly unzipped and imported. However, the lack of muzzling still is a concern. @tylerbenson PTAL 😃

**Review by tylerbenson — APPROVED** [review]
Consider squashing the changes so the jars don't end up in our history.

**Comment by kevinfwl** [comment]
> Consider squashing the changes so the jars don't end up in our history.

will do

**Review by richardstartin — APPROVED** [review]

**Pull request merged**